frontend/app/src/MetricsManager.ts

```diff
@@ -0,0 +1,376 @@
+  getCookie,
+  IGuestToHostMessage,
+  IMetricsEvent,
+  IS_DEV_ENV,
+  localStorageAvailable,
+  logAlways,
+  logError,
+  MetricsEvent,
+  SessionInfo,
+  setCookie,
+} from "@streamlit/lib"
+
+// Default metrics config fetched when none provided by host config endpoint
+export const DEFAULT_METRICS_CONFIG = "https://data.streamlit.io/metrics.json"
+
+/**
+ * The analytics is the Segment.io object. It is initialized in Segment.ts
+ * It is loaded with global scope (window.analytics) to integrate with the segment.io api
+ * @global
+ * */
+declare const analytics: any
+
+type EventName = "viewReport" | "updateReport" | "pageProfile" | "menuClick"
+type Event = [EventName, Partial<IMetricsEvent>]
+
+export class MetricsManager {
+  /** The app's SessionInfo instance. */
+  private readonly sessionInfo: SessionInfo
+
+  private initialized = false
+
+  /**
+   * Whether to send metrics to the server.
+   */
+  private actuallySendMetrics = false
+
+  /**
+   * The URL to which metrics are sent.
+   */
+  private metricsUrl: string | undefined = undefined
+
+  /**
+   * Function to send a message to the host via postMessage communication
+   */
+  private sendMessageToHost: (message: IGuestToHostMessage) => void = () => {}
+
+  /**
+   * The anonymous ID of the user.
+   */
+  private anonymousId = ""
+
+  /**
+   * Queue of metrics events that were enqueued before this MetricsManager was
+   * initialized.
+   */
+  private pendingEvents: Event[] = []
+
+  /**
+   * App hash uniquely identifies "projects" so we can tell
+   * how many projects are being created with Streamlit while still keeping
+   * possibly-sensitive info like the mainScriptPath outside of our metrics
+   * services.
+   */
+  private appHash = "Not initialized"
+
+  private metadata: DeployedAppMetadata = {}
+
+  public constructor(sessionInfo: SessionInfo) {
+    this.sessionInfo = sessionInfo
+  }
+
+  public initialize({
+    gatherUsageStats,
+    sendMessageToHost,
+  }: {
+    gatherUsageStats: boolean
+    sendMessageToHost: (message: IGuestToHostMessage) => void
+  }): void {
+    this.initialized = true
+    this.sendMessageToHost = sendMessageToHost
+    // Handle if the user or the host has disabled metrics
+    this.actuallySendMetrics = gatherUsageStats && this.metricsUrl !== "off"
+    this.getAnonymousId()
+
+    // Trigger fallback to fetch default metrics config if not provided by host
+    if (this.actuallySendMetrics && !this.metricsUrl) {
+      this.requestDefaultMetricsConfig()
+
+      // If metricsUrl still undefined, deactivate metrics
+      if (!this.metricsUrl) {
+        logError("Undefined metrics config - deactivating metrics tracking.")
+        this.actuallySendMetrics = false
+      }
+    }
+
+    if (this.actuallySendMetrics) {
+      // Segment will not initialize if this is rendered with SSR
+      initializeSegment()
+      this.sendPendingEvents()
+    }
+
+    logAlways("Gather usage stats: ", this.actuallySendMetrics)
+  }
```

Comment:
I think the `initialize` needs to be async and await the `requestDefaultMetricsConfig` method. Otherwise, the metrics gathering might be deactivated because the async function has not yet finished execution. And it would probably be good to move the `this.initialized = true` to the end, so that it can initialize async and collect pending events at the same time.

```typescript
  public async initialize({
    gatherUsageStats,
    sendMessageToHost,
  }: {
    gatherUsageStats: boolean
    sendMessageToHost: (message: IGuestToHostMessage) => void
  }): Promise<void> {
    this.sendMessageToHost = sendMessageToHost
    // Handle if the user or the host has disabled metrics
    this.actuallySendMetrics = gatherUsageStats && this.metricsUrl !== "off"
    this.getAnonymousId()

    // Trigger fallback to fetch default metrics config if not provided by host
    if (this.actuallySendMetrics && !this.metricsUrl) {
      await this.requestDefaultMetricsConfig()

      // If metricsUrl still undefined, deactivate metrics
      if (!this.metricsUrl) {
        logError("Undefined metrics config - deactivating metrics tracking.")
        this.actuallySendMetrics = false
      }
    }

    if (this.actuallySendMetrics) {
      // Segment will not initialize if this is rendered with SSR
      initializeSegment()
      this.sendPendingEvents()
    }

    logAlways("Gather usage stats: ", this.actuallySendMetrics)
    this.initialized = true
  }
```